File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Unit/Monster/ViewModel/TheBottom/SlimeBoss_ViewModel.ts
import { Mathf } from "../../../../../../../Core/Module/Math/Mathf";
import { AnimateSlowAttackAction } from "../../../../Action/Animations/AnimateSlowAttackAction";
import { ShakeScreenAction } from "../../../../Action/Animations/ShakeScreenAction";
import { ShoutAction } from "../../../../Action/Animations/ShoutAction";
import VFXAction from "../../../../Action/Animations/VFXAction";
import DamageAction from "../../../../Action/Common/DamageAction";
import { DoFuncAction } from "../../../../Action/Common/DoFuncAction";
import { MakeTempCardInDiscardAction } from "../../../../Action/Common/MakeTempCardInDiscardAction";
import { SetMoveAction } from "../../../../Action/Common/SetMoveAction";
import { SpawnMonsterAction } from "../../../../Action/Common/SpawnMonsterAction";
import { SuicideAction } from "../../../../Action/Common/SuicideAction";
import { HideHealthBarAction } from "../../../../Action/Utility/HideHealthBarAction";
import { SFXAction } from "../../../../Action/Utility/SFXAction";
import { TextAboveCreatureAction, TextType } from "../../../../Action/Utility/TextAboveCreatureAction";
import { WaitAction } from "../../../../Action/Utility/WaitAction";
import { SoundMaster } from "../../../../Audio/SoundMaster";
import { Slimed } from "../../../../Cards/Status/Slimed";
import { AttackEffect } from "../../../../DataDefine/AttackEffect";
import Color from "../../../../DataDefine/Color";
import DamageInfo from "../../../../DataDefine/DamageInfo";
import DungeonManager from "../../../../DungeonManager";
import { WeightyImpactEffect } from "../../../../Effect/Combat/WeightyImpactEffect";
import { ShakeIntensity, ShakeDur } from "../../../../Effect/ShakeScreen";
import { LocalizationProperty } from "../../../../Gen/DB/Localization";
import { AbstractMonster } from "../../AbstractMonster";
import { Intent } from "../../Intent";
import { AcidSlime_L_Model } from "../../Model/TheBottom/AcidSlime_L_Model";
import { SpikeSlime_L_Model } from "../../Model/TheBottom/SpikeSlime_L_Model";
import { AcidSlime_L_ViewModel } from "./AcidSlime_L_ViewModel";

export class SpikeSlime_L_ViewModel extends AbstractMonster {
    Initialize() {
        if (DungeonManager.Inst.AdvanceLevel >= 2) {
            this.DamageInfo.push(new DamageInfo(this, 10));
            this.DamageInfo.push(new DamageInfo(this, 38));
        } else {
            this.DamageInfo.push(new DamageInfo(this, 9));
            this.DamageInfo.push(new DamageInfo(this, 35));
        }

        this.SetAnimation("Idle", true);
    }

    Damage(info: DamageInfo) {
        if (!this.IsDying && this.Health <= this.MaxHealth / 2 && this.MoveIndex != 3) {
            this.SetMove(3, Intent.UNKNOWN, {moveName: "分裂"});
            this.AddToBot(new TextAboveCreatureAction(this, TextType.INTERRUPTED));
        }
    }

    GetMove(num: number) {
        if(DungeonManager.Inst.CurrentRoom.Round == 1){
            this.SetMove(4,Intent.STRONG_DEBUFF,{moveName:LocalizationProperty.Read(this.Info.Name + "行动4")});
        }
    }

    private PlaySfx(){
        let roll = Mathf.Floor(Mathf.RandomRange(0,2));
        if(roll == 0){
            this.AddToBot(new SFXAction("VO_SLIMEBOSS_1A"));
        } else if(roll == 1){
            this.AddToBot(new SFXAction("VO_SLIMEBOSS_1B"));
        } 
    }

    TakeTurn() {
        switch (this.MoveIndex) {
            case 4: {
                this.AddToBot(new AnimateSlowAttackAction(this));
                this.AddToBot(new SFXAction("MONSTER_SLIME_ATTACK"));
                if(DungeonManager.Inst.AdvanceLevel >= 19)
                    this.AddToBot(new MakeTempCardInDiscardAction(new Slimed(),5));
                else
                    this.AddToBot(new MakeTempCardInDiscardAction(new Slimed(),3));
                this.AddToBot(new SetMoveAction(this,2,Intent.UNKNOWN,{moveName:LocalizationProperty.Read(this.Info.Name + "行动2")}));
                break;
            }
            case 2:{
                this.PlaySfx();
                this.AddToBot(new ShoutAction(this,LocalizationProperty.Read(this.Info.Name + "对话1"),1,2))
                this.AddToBot(new ShakeScreenAction(0.3,ShakeIntensity.LOW, ShakeDur.LONG,{Horizontal: true,Vertical: false}));
                this.SetMove(1,Intent.ATTACK,{moveName:LocalizationProperty.Read(this.Info.Name + "行动1"),damage:this.DamageInfo[0]});
                break;
            }
            case 1: {
                this.AddToBot(new DoFuncAction(this.PlayJump,0.25));
                let playerHitBox = DungeonManager.Inst.Player.GetHitBox();
                this.AddToBot(new VFXAction(new WeightyImpactEffect(playerHitBox.CX,playerHitBox.CY,new Color(25,25,25,255)),0.8));
                this.AddToBot(new DamageAction(DungeonManager.Inst.Player, this.DamageInfo[1], AttackEffect.POISON));
                this.SetMove(4,Intent.STRONG_DEBUFF,{moveName:LocalizationProperty.Read(this.Info.Name + "行动4")});
                break;
            }
            case 3: {
                this.AddToBot(new AnimateSlowAttackAction(this));
                this.AddToBot(new HideHealthBarAction());
                this.AddToBot(new SuicideAction(this, false));
                this.AddToBot(new WaitAction(1000));
                this.AddToBot(new SFXAction("SLIME_SPLIT"));
                this.AddToBot(new SpawnMonsterAction(new AcidSlime_L_ViewModel(new AcidSlime_L_Model(this.X - 134, this.Y + Mathf.RandomRange(-4, 4))), false));
                this.AddToBot(new SpawnMonsterAction(new SpikeSlime_L_ViewModel(new SpikeSlime_L_Model(this.X + 134, this.Y + Mathf.RandomRange(-4, 4))), false));
                this.SetMove(3, Intent.UNKNOWN, {moveName: "分裂"});
            }
        }
    }


    Die() {
        super.Die();
        SoundMaster.PlayVoice("VO_SLIMEBOSS_2A");
        if(this.Health <= 0){
            this.PlayFastShake(5);
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Monster/ViewModel/TheBottom/SlimeBoss_ViewModel.cs
using Core.Utils;
using Core.Utils.Unity;
using ModTheSpire.Card.Status;
using ModTheSpire.DataDefine;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using ModTheSpire.Power.Buff;
using ModTheSpire.Power.Debuff;
using ModTheSpire.Unit.Monster.Helper;
using ModTheSpire.Unit.Monster.Model.TheBottom;
using Spine;
using Spine.Unity;
using TCG.Config;
using UnityEngine;

namespace ModTheSpire.Unit.Monster.ViewModel.TheBottom
{
    public class SlimeBoss_ViewModel : AbstractMonster
    {
        protected override void Initialize()
        {
            if(DungeonManager.Inst.AdvanceLevel >= 4) {
                this.DamageInfo.Add(new DamageInfo(this,10));
                this.DamageInfo.Add(new DamageInfo(this,38));

            }else{
                this.DamageInfo.Add(new DamageInfo(this,9));
                this.DamageInfo.Add(new DamageInfo(this,35));
            }
            var trackEntry = (TrackEntry)this.DisplayObject.SetAnimation("idle", true);
        }
        
        protected override void GetMove(int num)
        {
            if(DungeonManager.Inst.CurrentRoom.Round == 1)
            {
                this.SetMove(4, Intent.STRONG_DEBUFF, moveName: LocalizationProperty.Read(this.Info.Name + "行动4"));
            }
        }

        protected override void TakeTurn()
        {
switch (this.MoveIndex) {
            case 4: {
                this.AddToBot(new AnimateSlowAttackAction(this));
                this.AddToBot(new SFXAction(SoundMaster.MONSTER_SLIME_ATTACK));
                if(DungeonManager.Inst.AdvanceLevel >= 19)
                    this.AddToBot(new MakeTempCardInDiscardAction(new Slimed(),5));
                else
                    this.AddToBot(new MakeTempCardInDiscardAction(new Slimed(),3));
                SetMove(2,Intent.UNKNOWN,moveName:LocalizationProperty.Read(this.Info.Name + "行动2"));
                break;
            }
            case 2:{
                this.PlaySfx();
                this.AddToBot(new ShoutAction(this, LocalizationProperty.Read(this.Info.Name + "对话1"), 1, 2));
                this.AddToBot(new DoFuncAction(()=>CameraUtils.Shake(ShakeVibrato.LOW, ShakeDur.LONG, new Vector3(0.1f,0)),300));
                this.SetMove(1, Intent.ATTACK, this.DamageInfo[0], moveName: LocalizationProperty.Read(this.Info.Name + "行动1"));
                break;
            }
            case 1: {
                this.AddToBot(new DoFuncAction(this.DisplayObject.PlayJump,250));
                var playerHitBox = DungeonManager.Inst.Player.DisplayObject.GetHitBox();
                this.AddToBot(new VFXAction(new WeightyImpactEffect(playerHitBox.center.x,playerHitBox.center.y,new Color(25,25,25,255)),800));
                this.AddToBot(new DamageAction(DungeonManager.Inst.Player, this.DamageInfo[1], AttackEffect.POISON));
                this.SetMove(4,Intent.STRONG_DEBUFF,moveName:LocalizationProperty.Read(this.Info.Name + "行动4"));
                break;
            }
            case 3: {
                this.AddToBot(new AnimateSlowAttackAction(this));
                this.AddToBot(new HideHealthBarAction(this));
                this.AddToBot(new SuicideAction(this, false));
                this.AddToBot(new WaitAction(1000));
                this.AddToBot(new SFXAction(SoundMaster.SLIME_SPLIT));
                this.AddToBot(new SpawnMonsterAction(
                    new AcidSlime_L_ViewModel(new AcidSlime_L_Model(new Vector2(this.DisplayObject.X - 134,
                        this.DisplayObject.Y + Random.Range(-4, 5)))), false));
                this.AddToBot(new SpawnMonsterAction(
                    new SpikeSlime_L_ViewModel(new SpikeSlime_L_Model(new Vector2(this.DisplayObject.X + 134,
                        this.DisplayObject.Y + Random.Range(-4, 5)))), false));
                this.SetMove(3, Intent.UNKNOWN, moveName: "分裂");
                break;
            }
        }
        }
        private void PlaySfx()
        {
            int roll = Random.Range(0, 2);
            DungeonManager.ActionManager.AddToBottom(new SFXAction(roll switch
            {
                0 => SoundMaster.VO_SLIMEBOSS_1A,
                _ => SoundMaster.VO_SLIMEBOSS_1B,
            }));
        }

        protected override void Die()
        {
            base.Die();
            SoundMaster.PlayVoice(SoundMaster.VO_SLIMEBOSS_2A);
            if(this.Health <= 0){
                this.DisplayObject.PlayFastShake(5);
            }
        }

        public SlimeBoss_ViewModel(MonsterInfo info) : base(info)
        {
        }
    }
}